File: kotlin/src/test/kotlin/com/pragmastat/EstimatorsTest.kt
package com.pragmastat

import org.junit.jupiter.api.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue
import kotlin.test.assertFailsWith
import kotlin.math.abs

class EstimatorsTest {
    
    private val epsilon = 1e-10
    
    private fun assertClose(expected: Double, actual: Double, tolerance: Double = epsilon) {
        assertTrue(abs(expected - actual) < tolerance, 
            "Expected $expected but got $actual (difference: ${abs(expected - actual)})")
    }
    
    @Test
    fun testMedian() {
        assertFailsWith<IllegalArgumentException> { median(emptyList()) }
        assertEquals(1.0, median(listOf(1.0)))
        assertEquals(1.5, median(listOf(1.0, 2.0)))
        assertEquals(2.0, median(listOf(1.0, 2.0, 3.0)))
        assertEquals(2.5, median(listOf(1.0, 2.0, 3.0, 4.0)))
    }
    
    @Test
    fun testCenter() {
        assertFailsWith<IllegalArgumentException> { center(emptyList()) }
        assertEquals(1.0, center(listOf(1.0)))
        assertEquals(2.0, center(listOf(1.0, 3.0)))
        assertClose(2.0, center(listOf(1.0, 2.0, 3.0)))
        
        // Test with known values
        val x = listOf(10.0, 20.0)
        assertClose(15.0, center(x))
    }
    
    @Test
    fun testSpread() {
        assertFailsWith<IllegalArgumentException> { spread(emptyList()) }
        assertEquals(0.0, spread(listOf(1.0)))
        assertEquals(2.0, spread(listOf(1.0, 3.0)))
        
        // Test with known values
        val x = listOf(10.0, 20.0)
        assertClose(10.0, spread(x))
    }
    
    @Test
    fun testRelSpread() {
        assertEquals(0.0, relSpread(listOf(2.0)))
        assertFailsWith<IllegalArgumentException> { relSpread(listOf(0.0, 0.0)) }

        // Test with known values
        val x = listOf(10.0, 20.0)
        assertClose(10.0 / 15.0, relSpread(x))
    }

    @Test
    fun testShift() {
        assertFailsWith<IllegalArgumentException> { shift(emptyList(), listOf(1.0)) }
        assertFailsWith<IllegalArgumentException> { shift(listOf(1.0), emptyList()) }

        assertEquals(2.0, shift(listOf(3.0), listOf(1.0)))
        assertEquals(-2.0, shift(listOf(1.0), listOf(3.0)))

        // Test with multiple values
        val x = listOf(4.0, 5.0, 6.0)
        val y = listOf(1.0, 2.0, 3.0)
        assertClose(3.0, shift(x, y))
    }

    @Test
    fun testRatio() {
        assertFailsWith<IllegalArgumentException> { ratio(emptyList(), listOf(1.0)) }
        assertFailsWith<IllegalArgumentException> { ratio(listOf(1.0), emptyList()) }

        assertEquals(2.0, ratio(listOf(4.0), listOf(2.0)))
        assertEquals(0.5, ratio(listOf(2.0), listOf(4.0)))

        // Test with zero in denominator
        assertFailsWith<IllegalArgumentException> { ratio(listOf(1.0, 2.0), listOf(0.0)) }

        // Test with multiple values
        val x = listOf(4.0, 6.0, 8.0)
        val y = listOf(2.0, 3.0, 4.0)
        assertClose(2.0, ratio(x, y))
    }

    @Test
    fun testAvgSpread() {
        assertFailsWith<IllegalArgumentException> { avgSpread(emptyList(), emptyList()) }

        // Test with identical samples
        val x = listOf(1.0, 2.0, 3.0)
        val y = listOf(1.0, 2.0, 3.0)
        val spreadX = spread(x)
        assertClose(spreadX, avgSpread(x, y))

        // Test with different samples
        val a = listOf(1.0, 3.0)
        val b = listOf(2.0, 6.0)
        val spreadA = spread(a)
        val spreadB = spread(b)
        val expected = (2.0 * spreadA + 2.0 * spreadB) / 4.0
        assertClose(expected, avgSpread(a, b))
    }

    @Test
    fun testDisparity() {
        assertFailsWith<IllegalArgumentException> { disparity(emptyList(), emptyList()) }

        // Test with zero spread
        val x = listOf(1.0)
        val y = listOf(2.0)
        assertEquals(Double.POSITIVE_INFINITY, disparity(x, y))

        // Test normal case
        val a = listOf(1.0, 2.0, 3.0)
        val b = listOf(4.0, 5.0, 6.0)
        val shiftVal = shift(a, b)
        val spreadVal = avgSpread(a, b)
        assertClose(shiftVal / spreadVal, disparity(a, b))
    }
}

File: go/estimators_test.go
package pragmastat

import (
	"math"
	"testing"
)

// floatEquals checks if two float64 values are approximately equal
func floatEquals(a, b, epsilon float64) bool {
	if math.IsInf(a, 1) && math.IsInf(b, 1) {
		return true
	}
	if math.IsInf(a, -1) && math.IsInf(b, -1) {
		return true
	}
	if math.IsNaN(a) && math.IsNaN(b) {
		return true
	}
	return math.Abs(a-b) < epsilon
}

func TestMedian(t *testing.T) {
	tests := []struct {
		name     string
		input    []float64
		expected float64
	}{
		{"empty", []float64{}, 0.0},
		{"single", []float64{1.0}, 1.0},
		{"two elements", []float64{1.0, 2.0}, 1.5},
		{"three elements", []float64{1.0, 2.0, 3.0}, 2.0},
		{"four elements", []float64{1.0, 2.0, 3.0, 4.0}, 2.5},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if tt.name == "empty" {
				_, err := median(tt.input)
				if err == nil {
					t.Errorf("median(%v) expected error, got nil", tt.input)
				}
			} else {
				result, err := median(tt.input)
				if err != nil {
					t.Errorf("median(%v) unexpected error: %v", tt.input, err)
				}
				if !floatEquals(result, tt.expected, 1e-10) {
					t.Errorf("median(%v) = %v, want %v", tt.input, result, tt.expected)
				}
			}
		})
	}
}

func TestCenter(t *testing.T) {
	tests := []struct {
		name     string
		input    []float64
		expected float64
	}{
		{"empty", []float64{}, 0.0},
		{"single", []float64{1.0}, 1.0},
		{"two elements", []float64{1.0, 3.0}, 2.0},
		{"three elements", []float64{1.0, 2.0, 3.0}, 2.0},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if tt.name == "empty" {
				_, err := Center(tt.input)
				if err == nil {
					t.Errorf("Center(%v) expected error, got nil", tt.input)
				}
			} else {
				result, err := Center(tt.input)
				if err != nil {
					t.Errorf("Center(%v) unexpected error: %v", tt.input, err)
				}
				if !floatEquals(result, tt.expected, 1e-10) {
					t.Errorf("Center(%v) = %v, want %v", tt.input, result, tt.expected)
				}
			}
		})
	}
}

func TestSpread(t *testing.T) {
	tests := []struct {
		name     string
		input    []float64
		expected float64
	}{
		{"empty", []float64{}, 0.0},
		{"single", []float64{1.0}, 0.0},
		{"two elements", []float64{1.0, 3.0}, 2.0},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if tt.name == "empty" {
				_, err := Spread(tt.input)
				if err == nil {
					t.Errorf("Spread(%v) expected error, got nil", tt.input)
				}
			} else {
				result, err := Spread(tt.input)
				if err != nil {
					t.Errorf("Spread(%v) unexpected error: %v", tt.input, err)
				}
				if !floatEquals(result, tt.expected, 1e-10) {
					t.Errorf("Spread(%v) = %v, want %v", tt.input, result, tt.expected)
				}
			}
		})
	}
}

func TestRelSpread(t *testing.T) {
	tests := []struct {
		name     string
		input    []float64
		expected float64
	}{
		{"single non-zero", []float64{2.0}, 0.0},
		{"zeros", []float64{0.0, 0.0}, math.Inf(1)},
		{"normal case", []float64{10.0, 20.0}, 10.0 / 15.0}, // Spread([10,20]) = 10, Center([10,20]) = 15, RelSpread = 10/15
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if tt.name == "zeros" {
				_, err := RelSpread(tt.input)
				if err == nil {
					t.Errorf("RelSpread(%v) expected error, got nil", tt.input)
				}
			} else {
				result, err := RelSpread(tt.input)
				if err != nil {
					t.Errorf("RelSpread(%v) unexpected error: %v", tt.input, err)
				}
				if !floatEquals(result, tt.expected, 1e-10) {
					t.Errorf("RelSpread(%v) = %v, want %v", tt.input, result, tt.expected)
				}
			}
		})
	}
}

func TestShift(t *testing.T) {
	tests := []struct {
		name     string
		x        []float64
		y        []float64
		expected float64
	}{
		{"empty x", []float64{}, []float64{1.0}, 0.0},
		{"empty y", []float64{1.0}, []float64{}, 0.0},
		{"single elements", []float64{3.0}, []float64{1.0}, 2.0},
		{"opposite", []float64{1.0}, []float64{3.0}, -2.0},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if tt.name == "empty x" || tt.name == "empty y" {
				_, err := Shift(tt.x, tt.y)
				if err == nil {
					t.Errorf("Shift(%v, %v) expected error, got nil", tt.x, tt.y)
				}
			} else {
				result, err := Shift(tt.x, tt.y)
				if err != nil {
					t.Errorf("Shift(%v, %v) unexpected error: %v", tt.x, tt.y, err)
				}
				if !floatEquals(result, tt.expected, 1e-10) {
					t.Errorf("Shift(%v, %v) = %v, want %v", tt.x, tt.y, result, tt.expected)
				}
			}
		})
	}
}

func TestRatio(t *testing.T) {
	tests := []struct {
		name     string
		x        []float64
		y        []float64
		expected float64
	}{
		{"empty x", []float64{}, []float64{1.0}, 1.0},
		{"empty y", []float64{1.0}, []float64{}, 1.0},
		{"single elements", []float64{4.0}, []float64{2.0}, 2.0},
		{"opposite", []float64{2.0}, []float64{4.0}, 0.5},
		{"zero divisor", []float64{2.0}, []float64{0.0}, 1.0},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if tt.name == "empty x" || tt.name == "empty y" || tt.name == "zero divisor" {
				_, err := Ratio(tt.x, tt.y)
				if err == nil {
					t.Errorf("Ratio(%v, %v) expected error, got nil", tt.x, tt.y)
				}
			} else {
				result, err := Ratio(tt.x, tt.y)
				if err != nil {
					t.Errorf("Ratio(%v, %v) unexpected error: %v", tt.x, tt.y, err)
				}
				if !floatEquals(result, tt.expected, 1e-10) {
					t.Errorf("Ratio(%v, %v) = %v, want %v", tt.x, tt.y, result, tt.expected)
				}
			}
		})
	}
}

func TestAvgSpread(t *testing.T) {
	tests := []struct {
		name     string
		x        []float64
		y        []float64
		expected float64
	}{
		{"both empty", []float64{}, []float64{}, 0.0},
		{"x empty", []float64{}, []float64{1.0, 3.0}, 2.0},
		{"y empty", []float64{1.0, 3.0}, []float64{}, 2.0},
		{"equal sizes", []float64{1.0, 3.0}, []float64{5.0, 9.0}, 3.0},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if tt.name == "both empty" || tt.name == "x empty" || tt.name == "y empty" {
				_, err := AvgSpread(tt.x, tt.y)
				if err == nil {
					t.Errorf("AvgSpread(%v, %v) expected error, got nil", tt.x, tt.y)
				}
			} else {
				result, err := AvgSpread(tt.x, tt.y)
				if err != nil {
					t.Errorf("AvgSpread(%v, %v) unexpected error: %v", tt.x, tt.y, err)
				}
				if !floatEquals(result, tt.expected, 1e-10) {
					t.Errorf("AvgSpread(%v, %v) = %v, want %v", tt.x, tt.y, result, tt.expected)
				}
			}
		})
	}
}

func TestDisparity(t *testing.T) {
	tests := []struct {
		name     string
		x        []float64
		y        []float64
		expected float64
	}{
		{"both empty", []float64{}, []float64{}, math.Inf(1)},
		{"no difference", []float64{1.0, 2.0}, []float64{1.0, 2.0}, 0.0},
		{"normal case", []float64{3.0, 5.0}, []float64{1.0, 3.0}, 1.0},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if tt.name == "both empty" {
				_, err := Disparity(tt.x, tt.y)
				if err == nil {
					t.Errorf("Disparity(%v, %v) expected error, got nil", tt.x, tt.y)
				}
			} else {
				result, err := Disparity(tt.x, tt.y)
				if err != nil {
					t.Errorf("Disparity(%v, %v) unexpected error: %v", tt.x, tt.y, err)
				}
				if !floatEquals(result, tt.expected, 1e-10) {
					t.Errorf("Disparity(%v, %v) = %v, want %v", tt.x, tt.y, result, tt.expected)
				}
			}
		})
	}
}
